src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs

```diff
@@ -71,25 +70,43 @@ public static List<ZipGenericExtraField> ParseExtraField(ReadOnlySpan<byte> extr
                 totalBytesConsumed += currBytesConsumed;
                 extraFields.Add(field);
             }
+            // It's possible for some ZIP files to contain extra data which isn't a well-formed extra field. zipalign does this, padding the extra data
+            // field with zeroes to align the start of the file data to an X-byte boundary. We need to preserve this data so that the recorded "extra data length"
+            // fields in the central directory and local file headers are valid.
+            // We need to account for this because zipalign is part of the build process for Android applications, and failing to do this will result in
+            // .NET for Android generating corrupted .apk files and failing to build.
+            trailingExtraFieldData = extraFieldData[totalBytesConsumed..];
 
             return extraFields;
         }
 
-        public static int TotalSize(List<ZipGenericExtraField> fields)
+        public static int TotalSize(List<ZipGenericExtraField>? fields, int trailingDataLength)
         {
-            int size = 0;
-            foreach (ZipGenericExtraField field in fields)
+            int size = trailingDataLength;
+
+            if (fields != null)
             {
-                size += field.Size + SizeOfHeader; //size is only size of data
+                foreach (ZipGenericExtraField field in fields)
+                {
+                    size += field.Size + SizeOfHeader; //size is only size of data
+                }
             }
             return size;
         }
 
-        public static void WriteAllBlocks(List<ZipGenericExtraField> fields, Stream stream)
+        public static void WriteAllBlocks(List<ZipGenericExtraField>? fields, ReadOnlySpan<byte> trailingExtraFieldData, Stream stream)
         {
-            foreach (ZipGenericExtraField field in fields)
+            if (fields != null)
+            {
+                foreach (ZipGenericExtraField field in fields)
+                {
+                    field.WriteBlock(stream);
+                }
+            }
+
+            if (!trailingExtraFieldData.IsEmpty)
             {
-                field.WriteBlock(stream);
+                stream.Write(trailingExtraFieldData);
```

Comment:
Do you think there could be a case with no fields but with extra trailing data? Do we have a test for that?

Reply:
Thanks @carlossanlop. There can be, and I've just added test cases which cover this. The structure of an extra field is a two-byte tag ID, a two-byte data length, then a variable number of data bytes - so when we read four bytes of zeroes, we're actually reading a zero-length field with a tag ID of zero. The 0-3 bytes which follow this would be considered trailing data.

I've added tests which cover 3, 4, 7, 8 and 15 bytes of trailing data. This data should be interpreted by ZipArchive as 0, 1 or 2 well-formed zero-length fields, followed by 0 or 3 bytes of trailing data.